transactions/read-only/parallel.md

```diff
@@ -0,0 +1,215 @@
+    B -->|no| A
+    A --> D[write_window_deadline passed?]
+    D -->|yes| F[read-only trx queue empty?]
+    D -->|no| A
+    F -->|yes| A
+    F -->|no| R
+    
+    R --> S[read-only trx queue empty?]
+    S -->|yes| A
+    S -->|no| R
+    R --> T[read_window_deadline passed?]
+    T -->|yes| A
+    T -->|no| R  
+```
+
+### Handling main thread functions that are not safe to read-only transaction execution in `read` Window
+Several options are considered to handle 
+- Drop the functions. This in not acceptable as it changes the behavior.
+- Re-post to the main thread. In `read` window, re-post a not-read-only-safe function back to the main thread. This is simple to implement but breaks the order of functions and waste time in popping and pushing. 
+- Modify appbase `execution_priority_queue` by adding a function type with value `read-only-safe` or `not-read-only-safe`. In `write` window, everything works as it is now. In `read` window, only `thread-safe` requests are dequeued and processed. This avoids introducing new queues and keeps the order of the request. But as the queue is a priority queue based on function priority, it is infeasible to incorporate the additional type into the priority in a single queue.
+- Separate appbase queue into two queues: `read-only-safe` or `not-read-only-safe`. The function type is added when a function is posted. 
+  - In `read` window, only functions in `read-only-safe` queue are executed.
+  - In `write` window, compare the priorities of the top functions in `read-only-safe` and `not-read-only-safe`. The one with higher priority is executed. If tied, three options are considered:
+    - `not-read-only-safe` function is favored 
+    - randomly pick one
+    - add a time attribute to the functions and the older one is picked. This keeps the original behavior. Even though at a cost of the extra time field and an extra comparison, this option seems best.
+
+### Read-only Transaction Priority Queue
+The queue is used to store read-only transactions during `write` window. To maintain the time order when a transaction is put back into the queue for the next round due to `red` window deadline, a priority queue based on the first time when a transaction is queued is used.
+```c++
+struct read_only_trx {
+   fc::time_point          initial_queued_time;
+   packed_transaction_ptr  trx;
+   next_func_t             next;
+};
+      
+struct read_only_trx_less {
+   bool operator() (const read_only_trx& a, const read_only_trx& b) {
+      // earlier queued time has higher priority
+      return a.initial_queued_time > b.initial_queued_time;
+   }
+};
+
+std::priority_queue<read_only_trx, std::deque<read_only_trx>, read_only_trx_less>  read_only_trx_queue;     
+```
+
+### Configuration Options
+
+- `read-only-num-threads`: the number of threads in read-only transaction execution thread pool. Default to `0`. If it is `0`, read-only transactions are executed on the main thread sequentially as they arrive
```

Comment:
> Default to `0`. If it is `0`, read-only transactions are executed on the main thread sequentially as they arrive

Does this mean that, by default, this new functionality is disabled? Is it really what we want?

Reply:
Good point. Not sure what's a good number.

Reply:
As this feature should normally not run on producer nodes, disabling it by default seems better.

Reply:
> As this feature should normally not run on producer nodes

So if not run on producer nodes, it will not improve the chain max TPS, right?

Reply:
It helps indirectly by speeding up API (and PTP if servicing RPCs) nodes.

Reply:
I don't see how this would change the max chain TPS, though.

Reply:
That is not the point of these changes. If anything, this will decrease max TPS of a single producer node. That is one thing we will need to test. We should verify that making the priority queue thread safe and other changes do not greatly decrease max TPS for a single producer node.

Reply:
That's fine. Is there a document that describes the point of this change then? If not, maybe this document should start with a rationale for the changes. I originally assumed (apparently incorrectly) that the intent was to increase the chain TPS number.